### PR TITLE
Plugins signing: UI information

### PR DESCRIPTION
**Pull request by dprokop**
Based on https://github.com/grafana/grafana/pull/28219

Adds few notices about unsigned plugins in UI:
1. Plugins list:
image

2. Add new datasource:
image

3. Datasource picker:
image

4. Viz picker:
image

cc @daniellee @oanamangiurea

### CODE REVIEW

File: public/app/features/plugins/PluginsErrorsInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  }, [loadPlugins]);
+
+  if (loading || errors.length === 0) {
+    return null;
+  }
+  return (
+    <InfoBox
+      aria-label={selectors.pages.PluginsList.signatureErrorNotice}
+      severity="warning"
+      urlTitle="Read more about plugins signing"
```
Comment (torkelo) [inline public/app/features/plugins/PluginsErrorsInfo.tsx:46]:
```suggestion
      urlTitle="Read more about plugin signing"
```

making this singular reads a bit better

**Review by torkelo — APPROVED** [review]
Looks good!

File: public/app/features/plugins/PluginPage.tsx [hunk 2]
```diff
@@ -293,13 +298,48 @@ class PluginPage extends PureComponent<Props, State> {
...
+    if (plugin.meta.signature === PluginSignatureStatus.internal) {
+      return null;
+    }
+
+    return (
+      <InfoBox
+        aria-label={selectors.pages.PluginPage.signatureInfo}
+        severity={plugin.meta.signature !== PluginSignatureStatus.valid ? 'warning' : 'info'}
+        urlTitle="Read more about plugins signing"
+        url="TODO"
```
Comment (dprokop) [inline public/app/features/plugins/PluginPage.tsx:317]: @daniellee do we have a doc about that already?

File: packages/grafana-data/src/types/plugin.ts [hunk 3]
```diff
@@ -16,11 +16,23 @@ export enum PluginType {
 }
 
 export enum PluginSignatureStatus {
+  unsigned = 'unsigned', // no MANIFEST file
   internal = 'internal', // core plugin, no signature
   valid = 'valid', // signed and accurate MANIFEST
   invalid = 'invalid', // invalid signature
   modified = 'modified', // valid signature, but content mismatch
-  unsigned = 'unsigned', // no MANIFEST file
+  missing = 'missing', // missing signature file
```
Comment (wbrowne) [inline packages/grafana-data/src/types/plugin.ts:24]: Do we need `missing` here as it's technically the same as `unsigned`?
>// no MANIFEST file

>// missing signature file

File: packages/grafana-data/src/types/plugin.ts [hunk 4]
```diff
@@ -16,11 +16,23 @@ export enum PluginType {
...
+}
+
+export enum PluginErrorCode {
```
Comment (wbrowne) [inline packages/grafana-data/src/types/plugin.ts:27]: Is it confusing having signature related statuses in both `PluginSignatureStatus` and `PluginErrorCode`? Should we maybe rename `PluginErrorCode` as `BackendPluginErrorCode` instead?

File: packages/grafana-ui/src/components/Alert/Alert.tsx [hunk 5]
```diff
@@ -76,21 +77,11 @@ export const Alert: FC<Props> = ({
...
-    `,
-    success: css`
-      background: linear-gradient(100deg, ${greenBase}, ${greenShade});
-    `,
-  };
+  const { white } = theme.palette;
+  const severityColors = getColorsFromSeverity(severity, theme);
+  const background = css`
+    background: linear-gradient(90deg, ${severityColors[0]}, ${severityColors[0]});
+  `;
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Alert/Alert.tsx:84]: Very nice!

File: packages/grafana-ui/src/components/Badge/Badge.tsx [hunk 6]
```diff
@@ -1,28 +1,28 @@
...
 import tinycolor from 'tinycolor2';
-import { css } from 'emotion';
-import { HorizontalGroup } from '..';
+import { css, cx } from 'emotion';
+import { HorizontalGroup } from '../Layout/Layout';
 
 export type BadgeColor = 'blue' | 'red' | 'green' | 'orange' | 'purple';
 
-export interface BadgeProps {
+export interface BadgeProps extends HTMLAttributes<HTMLDivElement> {
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Badge/Badge.tsx:14]: Yes!

File: public/app/core/components/Select/DataSourcePicker.tsx [hunk 7]
```diff
@@ -85,6 +88,21 @@ export class DataSourcePicker extends PureComponent<Props> {
...
+              o.meta.signature !== PluginSignatureStatus.internal &&
+              o !== value
+            ) {
+              return (
+                <HorizontalGroup align="center" justify="space-between">
+                  <span>{o.label}</span> <PluginSignatureBadge status={o.meta.signature} />
+                </HorizontalGroup>
+              );
+            }
+            return o.label || '';
```
Comment (hugohaggmark) [inline public/app/core/components/Select/DataSourcePicker.tsx:104]: nit: extract to function?

File: public/app/features/dashboard/panel_editor/VizTypePickerPlugin.tsx [hunk 8]
```diff
@@ -135,6 +136,13 @@ interface PanelPluginBadgeProps {
...
+  if (
+    plugin.signature &&
+    plugin.signature !== PluginSignatureStatus.valid &&
+    plugin.signature !== PluginSignatureStatus.internal
+  ) {
+    return <PluginSignatureBadge status={plugin.signature} />;
+  }
   if (plugin.state !== PluginState.deprecated && plugin.state !== PluginState.alpha) {
     return null;
   }
```
Comment (hugohaggmark) [inline public/app/features/dashboard/panel_editor/VizTypePickerPlugin.tsx:148]: nit: what if we had this logic inside a container element that wraps PluginSignatureBadge?

**Review by hugohaggmark — APPROVED** [review]
Fantastic work ⭐  left some minor nits for consideration.

File: packages/grafana-ui/src/components/InfoBox/InfoBox.tsx [hunk 9]
```diff
@@ -7,13 +7,16 @@ import { IconButton } from '../IconButton/IconButton';
 import { HorizontalGroup } from '../Layout/Layout';
 import panelArtDark from './panelArt_dark.svg';
 import panelArtLight from './panelArt_light.svg';
+import { AlertVariant } from '..';
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/InfoBox/InfoBox.tsx:10]: Does this shorthand import work in builds as well? I remember there were some issues with those but could be wrong.

File: public/app/features/dashboard/panel_editor/VizTypePickerPlugin.tsx [hunk 10]
```diff
@@ -135,6 +136,13 @@ interface PanelPluginBadgeProps {
 const PanelPluginBadge: React.FC<PanelPluginBadgeProps> = ({ plugin }) => {
   const display = getPanelStateBadgeDisplayModel(plugin);
 
+  if (
+    plugin.signature &&
+    plugin.signature !== PluginSignatureStatus.valid &&
+    plugin.signature !== PluginSignatureStatus.internal
```
Comment (Clarity-89) [inline public/app/features/dashboard/panel_editor/VizTypePickerPlugin.tsx:142]: Similar logic to the one in `DataSourcePicker.tsx` on lines 93-95. I'd consider extracting this into a function.

File: public/app/features/plugins/PluginListPage.test.tsx [hunk 11]
```diff
@@ -1,11 +1,27 @@
...
+import { render, screen, waitFor } from '@testing-library/react';
+import { selectors } from '@grafana/e2e-selectors';
+import { Provider } from 'react-redux';
+import { configureStore } from '../../store/configureStore';
+import { afterEach } from '../../../test/lib/common';
+
+let errorsReturnMock: any = [];
+
+jest.mock('@grafana/runtime', () => ({
+  ...((jest.requireActual('@grafana/runtime') as unknown) as object),
```
Comment (Clarity-89) [inline public/app/features/plugins/PluginListPage.test.tsx:15]:
```suggestion
  ...(jest.requireActual('@grafana/runtime') as object),
```

File: public/app/features/plugins/PluginListPage.tsx [hunk 12]
```diff
@@ -20,40 +23,49 @@ export interface Props {
...
+export const PluginListPage: React.FC<Props> = ({
+  hasFetched,
+  navModel,
+  plugins,
+  setPluginsSearchQuery,
+  searchQuery,
+  loadPlugins,
+}) => {
+  useAsync(async () => {
+    return await loadPlugins();
```
Comment (Clarity-89) [inline public/app/features/plugins/PluginListPage.tsx:35]: I think `await` is redundant here.

File: public/app/features/plugins/PluginListPage.tsx [hunk 13]
```diff
@@ -20,40 +23,49 @@ export interface Props {
...
-      </Page>
-    );
-  }
-}
+          <PluginsErrorsInfo>
+            <>
+              <br />
+              <p>
+                Note that <strong>unsigned front-end datasource and panel plugins</strong> are still usable, but this is
+                a subject of change in the upcoming releases of Grafana
```
Comment (Clarity-89) [inline public/app/features/plugins/PluginListPage.tsx:59]:
```suggestion
                subject to change in the upcoming releases of Grafana
```

File: public/app/features/datasources/NewDataSourcePage.tsx [hunk 14]
```diff
@@ -98,6 +99,17 @@ class NewDataSourcePage extends PureComponent<Props> {
             <div className="page-action-bar__spacer" />
             <LinkButton href="datasources">Cancel</LinkButton>
           </div>
+          {!searchQuery && (
+            <PluginsErrorsInfo>
+              <>
+                <br />
+                <p>
+                  Note that <strong>unsigned front-end datasource plugins</strong> are still usable, but this is a
+                  subject of change in the upcoming releases of Grafana
```
Comment (Clarity-89) [inline public/app/features/datasources/NewDataSourcePage.tsx:108]:
```suggestion
                  Note that <strong>unsigned front-end datasource plugins</strong> are still usable, but this is
                  subject to change in the upcoming releases of Grafana
```

File: public/app/features/plugins/PluginListPage.tsx (hunk 13 above)
Comment (Clarity-89) [reply public/app/features/plugins/PluginListPage.tsx:59]: Also this is a second time this text/component is used, so maybe it's better to extract this?

File: public/app/features/plugins/PluginSignatureBadge.tsx [hunk 15]
```diff
@@ -1,16 +1,37 @@
...
+      text={display.text}
+      color={display.color as any}
+      icon={display.icon}
+      tooltip={display.tooltip}
+      {...otherProps}
+    />
+  );
 };
 
+export function mapPluginErrorCodeToSignatureStatus(code: PluginErrorCode) {
```
Comment (Clarity-89) [inline public/app/features/plugins/PluginSignatureBadge.tsx:22]: Could also use `Map` for this.

File: public/app/features/plugins/PluginListPage.tsx (hunk 13 above)
Comment (dprokop) [reply public/app/features/plugins/PluginListPage.tsx:59]: it's a different text ;)

Comment (Clarity-89) [reply public/app/features/plugins/PluginListPage.tsx:59]: Ah, you're right, noticed the subtle difference 🙈

File: packages/grafana-data/src/types/plugin.ts (hunk 4 above)
Comment (dprokop) [reply packages/grafana-data/src/types/plugin.ts:27]: My thinking is that `PluginErrorCode` may represent other types of plugin errors, not only the signature ones

File: public/app/features/plugins/PluginListPage.tsx (hunk 12 above)
Comment (dprokop) [reply public/app/features/plugins/PluginListPage.tsx:35]: yes!

File: public/app/features/dashboard/panel_editor/VizTypePickerPlugin.tsx (hunk 8 above)
Comment (dprokop) [reply public/app/features/dashboard/panel_editor/VizTypePickerPlugin.tsx:148]: Yeah, this is worth considering but the changes we are introducing here are going to change soon, so I would skip this for now.

File: public/app/features/plugins/PluginPage.tsx (hunk 2 above)
Comment (oddlittlebird) [reply public/app/features/plugins/PluginPage.tsx:317]: @marcusolsson ^

Comment (marcusolsson) [reply public/app/features/plugins/PluginPage.tsx:317]: https://grafana.com/docs/grafana/latest/plugins/plugin-signature-verification/

**Review by achatterjee-grafana — APPROVED** [review]
Code changes and doc snippets looks good.

File: public/app/features/plugins/PluginsErrorsInfo.tsx [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+  if (loading || errors.length === 0) {
+    return null;
+  }
+  return (
+    <InfoBox
+      aria-label={selectors.pages.PluginsList.signatureErrorNotice}
+      severity="warning"
+      urlTitle="Read more about plugin signing"
+      url="TODO"
```
Comment (dprokop) [inline public/app/features/plugins/PluginsErrorsInfo.tsx:47]:
```suggestion
      url="https://grafana.com/docs/grafana/latest/plugins/plugin-signature-verification/"
```

File: public/app/features/plugins/PluginPage.tsx (hunk 2 above)
Comment (dprokop) [inline public/app/features/plugins/PluginPage.tsx:317]:
```suggestion
        url="https://grafana.com/docs/grafana/latest/plugins/plugin-signature-verification/"
```

File: pkg/plugins/models.go [hunk 17]
```diff
@@ -197,3 +197,8 @@ func NewEnabledPlugins() EnabledPlugins {
 		Apps:        make([]*AppPlugin, 0),
 	}
 }
+
+type PluginErrorInfo struct {
+	ErrorCode string `json:"errorCode"`
+	PluginID  string `json:"pluginId,omitempty"`
+}
```
Comment (dprokop) [inline pkg/plugins/models.go:204]: @wbrowne can you have a look at these go files? Has there been any changes in your PR when you merged? I don't think these changes should be here as I haven't touched any backend code.

Comment (wbrowne) [reply pkg/plugins/models.go:204]: Yeah exactly. These changes shouldn't be here. All my stuff is now merged to master - so you should just need to be merge/rebase and see all my changes

Comment (dprokop) [reply pkg/plugins/models.go:204]: eh, I guess it's a squash merge issue as this is now synced with master and the changes are still here

**Review by wbrowne — APPROVED** [review]
LGTM 🕺 

Frontend docs failing though 😢 
>API Extractor warnings/errors 1078 exceeded 1077 so failing build.

**Comment by wbrowne** [comment]
@mckn Looks like the `WARNINGS_COUNT_LIMIT` has exceeded by one 😅  Should this be just bumped slightly or should there be a better magic number here?

**Comment by mckn** [comment]
> @mckn Looks like the `WARNINGS_COUNT_LIMIT` has exceeded by one 😅 Should this be just bumped slightly or should there be a better magic number here?

I think we need to address the underlying issue and add documentation for the newly added code :)

**Pull request merged**